.github/workflows/publish-algorithm-images.yaml

```diff
@@ -0,0 +1,99 @@
+      image: docker.io/kubeflowkatib/suggestion-hyperopt
+      dockerfile: cmd/suggestion/hyperopt/v1beta1/Dockerfile
+    secrets:
+      dockerhub_username: ${{ secrets.DOCKERHUB_USERNAME }}
+      dockerhub_token: ${{ secrets.DOCKERHUB_TOKEN }}
+
+  suggestion-chocolate:
+    name: Suggestion Chocolate
+    uses: kubeflow/katib/.github/workflows/template-publish-image.yaml@master
+    with:
+      image: docker.io/kubeflowkatib/suggestion-chocolate
+      dockerfile: cmd/suggestion/chocolate/v1beta1/Dockerfile
+    secrets:
+      dockerhub_username: ${{ secrets.DOCKERHUB_USERNAME }}
+      dockerhub_token: ${{ secrets.DOCKERHUB_TOKEN }}
+
+  suggestion-hyperband:
+    name: Suggestion Hyperband
+    uses: kubeflow/katib/.github/workflows/template-publish-image.yaml@master
+    with:
+      image: docker.io/kubeflowkatib/suggestion-hyperband
+      dockerfile: cmd/suggestion/hyperband/v1beta1/Dockerfile
+    secrets:
+      dockerhub_username: ${{ secrets.DOCKERHUB_USERNAME }}
+      dockerhub_token: ${{ secrets.DOCKERHUB_TOKEN }}
+
+  suggestion-skopt:
+    name: Suggestion Skopt
+    uses: kubeflow/katib/.github/workflows/template-publish-image.yaml@master
+    with:
+      image: docker.io/kubeflowkatib/suggestion-skopt
+      dockerfile: cmd/suggestion/skopt/v1beta1/Dockerfile
+    secrets:
+      dockerhub_username: ${{ secrets.DOCKERHUB_USERNAME }}
+      dockerhub_token: ${{ secrets.DOCKERHUB_TOKEN }}
+
+  suggestion-goptuna:
+    name: Suggestion Goptuna
+    uses: kubeflow/katib/.github/workflows/template-publish-image.yaml@master
+    with:
+      image: docker.io/kubeflowkatib/suggestion-goptuna
+      dockerfile: cmd/suggestion/goptuna/v1beta1/Dockerfile
+    secrets:
+      dockerhub_username: ${{ secrets.DOCKERHUB_USERNAME }}
+      dockerhub_token: ${{ secrets.DOCKERHUB_TOKEN }}
+
+  suggestion-optuna:
+    name: Suggestion Optuna
+    uses: kubeflow/katib/.github/workflows/template-publish-image.yaml@master
+    with:
+      image: docker.io/kubeflowkatib/suggestion-optuna
+      dockerfile: cmd/suggestion/optuna/v1beta1/Dockerfile
+    secrets:
+      dockerhub_username: ${{ secrets.DOCKERHUB_USERNAME }}
+      dockerhub_token: ${{ secrets.DOCKERHUB_TOKEN }}
+
+  suggestion-enas:
+    name: Suggestion ENAS
+    uses: kubeflow/katib/.github/workflows/template-publish-image.yaml@master
+    with:
+      image: docker.io/kubeflowkatib/suggestion-enas
+      dockerfile: cmd/suggestion/nas/enas/v1beta1/Dockerfile
+    secrets:
+      dockerhub_username: ${{ secrets.DOCKERHUB_USERNAME }}
+      dockerhub_token: ${{ secrets.DOCKERHUB_TOKEN }}
+
+  suggestion-darts:
+    name: Suggestion DARTS
+    uses: kubeflow/katib/.github/workflows/template-publish-image.yaml@master
+    with:
+      image: docker.io/kubeflowkatib/suggestion-darts
+      dockerfile: cmd/suggestion/nas/darts/v1beta1/Dockerfile
+    secrets:
+      dockerhub_username: ${{ secrets.DOCKERHUB_USERNAME }}
+      dockerhub_token: ${{ secrets.DOCKERHUB_TOKEN }}
```

Comment:
Should we use [`composite-run-steps`](https://github.com/actions/runner/blob/main/docs/adrs/0549-composite-run-steps.md) and [`matrix`](https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#jobsjob_idstrategymatrix) to further reuse action scripts?

```suggestion
jobs:
  # Suggestion Images.
  suggestion:
    runs-on: ubuntu-20.04
    strategy:
      matrix:
        suggestion-name:
          - "hyperopt"
          - "chocolate"
          - "hyperband"
          - "skopt"
          - "goptuna"
          - "optuna"
        basedir: ["suggestion"]
        include:
          - suggestion-name: "enas"
            basedir: "suggestion/nas"
          - suggestion-name: "darts"
            basedir: "suggestion/nas"
    steps:
      - name: Checkout
        uses: actions/checkout@v2
        with:
          fetch-depth: 1
      - name: Suggestion ${{ matrix.suggestion-name }}
        uses: ./.github/workflows/template-publish-image
        with:
          image: docker.io/kubeflowkatib/suggestion-${{ matrix.suggestion-name }}
          dockerfile: cmd/${{ matrix.basedir }}/${{ matrix.suggestion-name }}/v1beta1/Dockerfile
```

Reply:
In that approach each image will be built in a separate `step`, right ?
I use separate job to build each image in favour on this feature: https://github.com/github/roadmap/issues/271
Once it will be implemented we can easily restart the specific job that was failed (for example connection issue), instead of re-run the whole workflow.
What do you think @tenzen-y ?

Reply:
> In that approach each image will be built in a separate step, right ?

Maybe, I think my suggestion script will build in a separate `job` in the following.
Is the following different from the `job` you are expecting? @andreyvelich 

https://github.com/tenzen-y/katib/actions/runs/1506407092

<img width="1166" alt="スクリーンショット 2021-11-27 0 55 22" src="https://user-images.githubusercontent.com/68272500/143606154-ebcf7334-cab6-4a04-bfc7-4452d4c24ff7.png">

Reply:
As I can see it also creates the separate Job for each build, thank you for the testing.
Let me try to test it on my end also.

Reply:
@tenzen-y Do you know how to pass secret in steps once we apply reusable workflow on the step level ?
As I can see, secrets can be accessible only in job level: https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#jobsjob_idsecrets.

Reply:
If we use `composite action`, we can pass token and username using the global env field in the following.

- publish-algorithm-images.yaml

```yaml
env:
  dockerhub_username: ${{ secrets.DOCKERHUB_USERNAME }}
  dockerhub_token: ${{ secrets.DOCKERHUB_TOKEN }}
on: push

jobs:
  # Suggestion Images.
...
```

- action.yaml

```yaml
...
  using: "composite"
  steps:

    - name: Set Up Docker Buildx
      uses: docker/setup-buildx-action@v1

    - name: Docker Login
      uses: docker/login-action@v1
      with:
        username: ${{ env.dockerhub_username }}
        password: ${{ env.dockerhub_token }}
...
```

The following is the example that I actually run actions in my forked repository.

- [actions log](https://github.com/tenzen-y/katib/runs/4337719723?check_suite_focus=true#step:3:142)

```shell
...
Run docker/login-action@v1
  with:
    username: ***
    password: ***
    logout: true
  env:
    dockerhub_username: ***
    dockerhub_token: ***
Logging into Docker Hub...
Login Succeeded!
...
```

- Settings/Secrets

<img width="796" alt="スクリーンショット 2021-11-27 6 19 46" src="https://user-images.githubusercontent.com/68272500/143656048-f733c613-655d-434d-acbf-408163fd3227.png">

Reply:
Thank you for this information @tenzen-y!
I refactored the workflow using composite run.
To be consistent and easy to read, I use similar `matrix` structure for Core, Algorithms, and Trials workflow.
In that case, we don't need to create separate job for Early Stopping, Metrics Collector, etc.
Please let me know what do you think.

Reply:
It sounds good to me! @andreyvelich